website/blog/2024-02-29-StateFlow/index.mdx

```diff
@@ -0,0 +1,141 @@
+# Define the agents, the code is for illustration purposes and is not executable.
+initializer = autogen.UserProxyAgent(
+   name="Init"
+)
+coder = autogen.AssistantAgent(
+   name="Coder",
+   system_message="""You are the Coder. Write Python Code to retrieve papers from arxiv.."""
+)
+executor = autogen.UserProxyAgent(
+   name="Executor",
+   system_message="Executor. Execute the code written by the Coder and report the result.",
+)
+scientist = autogen.AssistantAgent(
+   name="Scientist",
+   system_message="""You are the Scientist. Please categorize papers after seeing their abstracts printed and create a markdown table with Domain, Title, Authors, Summary and Link. Return 'TERMINATE' in the end.""",
+)
+```
+
+In the Figure, we define a simple workflow for research with 4 states: Init, Retrieve, Reserach and End. Within each state, we will call different agents to perform the tasks.
+- Init: We use the initializer to start the workflow.
+- Retrieve: We will first call the coder to write code and then call the executor to execute the code.
+- Research: We will call the scientist to read the papers and write a summary.
+- End: We will end the workflow.
+
+
+Then we define a customized function to control the transition between states:
+```python
+def state_transition(last_speaker, groupchat):
+   messages = groupchat.messages
+
+
+   if last_speaker is initializer:
+       # init -> retrieve
+       return coder
+   elif last_speaker is coder:
+       # retrieve: action 1 -> action 2
+       return executor
+   elif last_speaker is executor:
+       if messages[-1]["content"] == "exitcode: 1":
+           # retrieve --(execution failed)--> retrieve
+           return coder
+       else:
+           # retrieve --(execution success)--> research
+           return scientist
+   elif last_speaker == "Scientist":
+       # research -> end
+       return None
+
+
+groupchat = autogen.GroupChat(
+   agents=[initializer, coder, executor, scientist],
+   messages=[],
+   max_round=20,
+   speaker_selection_method=state_transition,
+)
+```
+
+We recommend implementing the transition logic for each speaker in the customized function. In analogy to a state machine, a state transition function determines the next state based on the current state and input.
+Instead of returning an `Agent` class representing the next speaker, we can also return a string from `['auto', 'manual', 'random', 'round_robin']` to select a default method to use.
+For example, we can always default to the built-in `auto` method to employ a LLM-based group chat manager to select the next speaker.
+When returning `None`, the group chat will terminate.
+
+
+## For Further Reading
+* [StateFlow notebook](https://microsoft.github.io/autogen/docs/notebooks/agentchat_groupchat_stateflow/)
+* [GroupChat with Customized Speaker Selection notebook](https://microsoft.github.io/autogen/docs/notebooks/agentchat_groupchat_customized/)
+* [FSM Group Chat](/blog/2024/02/11/FSM-GroupChat/)
```

Comment:
The new blog doesn't discuss relations with FSM group chat, which could confuse readers. I'd suggest adding discussions.